ambry-server/src/main/java/com.github.ambry.server/VCRServer.java

```diff
@@ -0,0 +1,240 @@
+  private MetricRegistry registry = null;
+  private JmxReporter reporter = null;
+  private ConnectionPool connectionPool = null;
+  private final NotificationSystem notificationSystem;
+  private CloudDestinationFactory cloudDestinationFactory;
+  private SSLConfig sslConfig;
+  private VCRMetrics metrics = null;
+
+  /**
+   * VCRServer constructor.
+   * @param properties the config properties to use.
+   * @param clusterAgentsFactory the {@link ClusterAgentsFactory} to use.
+   * @param notificationSystem the {@link NotificationSystem} to use.
+   */
+  public VCRServer(VerifiableProperties properties, ClusterAgentsFactory clusterAgentsFactory,
+      NotificationSystem notificationSystem) {
+    this.properties = properties;
+    this.clusterAgentsFactory = clusterAgentsFactory;
+    this.notificationSystem = notificationSystem;
+  }
+
+  /** Test constructor */
+  VCRServer(VerifiableProperties properties, ClusterAgentsFactory clusterAgentsFactory,
+      NotificationSystem notificationSystem, CloudDestinationFactory cloudDestinationFactory,
+      VirtualReplicatorCluster virtualReplicatorCluster, SSLConfig sslConfig) {
+    this(properties, clusterAgentsFactory, notificationSystem);
+    this.cloudDestinationFactory = cloudDestinationFactory;
+    this.virtualReplicatorCluster = virtualReplicatorCluster;
+    this.sslConfig = sslConfig;
+  }
+
+  /**
+   * Start the VCR Server.
+   * @throws InstantiationException if an error was encountered during startup.
+   */
+  public void startup() throws InstantiationException {
+    try {
+      logger.info("starting");
+      clusterMap = clusterAgentsFactory.getClusterMap();
+      logger.info("Initialized clusterMap");
+      clusterParticipant = clusterAgentsFactory.getClusterParticipant();
+
+      logger.info("Setting up JMX.");
+      long startTime = SystemTime.getInstance().milliseconds();
+      registry = clusterMap.getMetricRegistry();
+      this.metrics = new VCRMetrics(registry);
+      reporter = JmxReporter.forRegistry(registry).build();
+      reporter.start();
+
+      logger.info("creating configs");
+      NetworkConfig networkConfig = new NetworkConfig(properties);
+      StoreConfig storeConfig = new StoreConfig(properties);
+      ServerConfig serverConfig = new ServerConfig(properties);
+      ReplicationConfig replicationConfig = new ReplicationConfig(properties);
+      CloudConfig cloudConfig = new CloudConfig(properties);
+      ConnectionPoolConfig connectionPoolConfig = new ConnectionPoolConfig(properties);
+      ClusterMapConfig clusterMapConfig = new ClusterMapConfig(properties);
+      if (sslConfig == null) {
+        sslConfig = new SSLConfig(properties);
+      }
+      // verify the configs
+      properties.verify();
+
+      // Note: using static cluster until Helix version implemented
+      if (virtualReplicatorCluster == null) {
+        virtualReplicatorCluster = new StaticVCRCluster(cloudConfig, clusterMapConfig, clusterMap);
```

Comment:
It's better to get `virtualReplicatorCluster` from config, just like `cloudDestinationFactory`.

Reply:
Agree, but do we have factory for it?